File: nix/nixpkgs/devlogs-lib/devlogs.go
package devlogs

import (
	"fmt"
	"os"
	"os/exec"
	"strings"
)

type Logger struct {
	component string
	window    string
}

func NewLogger(component string) *Logger {
	l := &Logger{component: component}
	if pane := os.Getenv("TMUX_PANE"); pane != "" {
		out, err := exec.Command("tmux", "display-message", "-t", pane, "-p", "#{window_index}").Output()
		if err == nil {
			l.window = strings.TrimSpace(string(out))
		}
	}
	return l
}

func (l *Logger) Debug(msg string) { l.log("debug", msg) }
func (l *Logger) Info(msg string)  { l.log("info", msg) }
func (l *Logger) Warn(msg string)  { l.log("warning", msg) }
func (l *Logger) Error(msg string) { l.log("err", msg) }

func (l *Logger) log(level, msg string) {
	tag := l.component
	if l.window != "" {
		tag = fmt.Sprintf("%s(@%s)", l.component, l.window)
	}
	formatted := fmt.Sprintf("[devlogs] %s %s: %s", strings.ToUpper(level), tag, msg)
	// Errors silently suppressed, matching Python/shell behaviour
	_ = exec.Command("logger", "-t", "devlogs", "-p", "user."+level, formatted).Run()
}

File: nix/nixpkgs/devlogs-lib/devlogs.py
"""Devlogs logging library — unified syslog logging for Python scripts."""

import contextlib
import logging
import os
import subprocess
from typing import Final

PRIORITY_MAP: Final = {
    logging.DEBUG: "user.info",  # macOS drops user.debug from history
    logging.INFO: "user.info",
    logging.WARNING: "user.warning",
    logging.ERROR: "user.err",
    logging.CRITICAL: "user.crit",
}


class SyslogHandler(logging.Handler):
    """Log via /usr/bin/logger for macOS unified logging."""

    def emit(self, record: logging.LogRecord) -> None:
        priority = PRIORITY_MAP.get(record.levelno, "user.info")
        msg = self.format(record)
        with contextlib.suppress(FileNotFoundError, subprocess.TimeoutExpired, OSError):
            subprocess.run(
                ["logger", "-t", "devlogs", "-p", priority, msg],
                timeout=2,
            )


def setup_logging(component: str = "", instance: str = "") -> logging.Logger:
    """Create a logger that writes to syslog in devlogs format.

    Args:
        component: Log component name. Falls back to DEVLOGS_COMPONENT env, then "unknown".
        instance: Instance identifier. Falls back to DEVLOGS_INSTANCE env, then "-".
    """
    if not component:
        component = os.environ.get("DEVLOGS_COMPONENT", "unknown")
    if not instance:
        instance = os.environ.get("DEVLOGS_INSTANCE", "-")

    window = os.environ.get("TARGET_WINDOW", "")
    tag = f"{component}{{{instance}}}"
    if window:
        tag = f"{component}{{{instance}}}(@{window})"
    logger = logging.getLogger(f"devlogs.{component}")
    logger.setLevel(logging.DEBUG)
    handler = SyslogHandler()
    handler.setFormatter(
        logging.Formatter(f"[devlogs] %(levelname)s {tag}: %(message)s")
    )
    logger.addHandler(handler)
    return logger
